matscipy/calculators/eam/precon.py

```diff
@@ -0,0 +1,204 @@
+                 r_cut=None, mu=None, mu_c=None, dim=3, c_stab=0.1,
+                 force_stab=False,
+                 reinitialize=False, array_convention='C',
+                 solver="auto", solve_tol=1e-9,
+                 apply_positions=True, apply_cell=True, logfile=None):
+        """Initialise an Hessian_EAM preconditioner with given parameters.
+
+        Args:
+            calculator: matscipy eam calculator
+                The calculator that calculates the hessian of the system.
+            nbr_of_evals: int
+                The number of evaluation after the matrix is calculated.
+            terms: list of strings
+                The list which terms of the analytic eam hessian should be
+                calculated.
+            r_cut, mu, c_stab, dim, recalc_mu, array_convention: see
+                precon.__init__()
+        """
+        super().__init__(r_cut=r_cut, mu=mu, mu_c=mu_c,
+                         dim=dim, c_stab=c_stab,
+                         force_stab=force_stab,
+                         reinitialize=reinitialize,
+                         array_convention=array_convention,
+                         solver=solver, solve_tol=solve_tol,
+                         apply_positions=apply_positions,
+                         apply_cell=apply_cell,
+                         logfile=logfile)
+        
+        self.calculator = calculator
+        self.terms = terms
+        self.nbr_of_evals = nbr_of_evals
+        self.ctr_of_evals = 0
+
+    def make_precon(self, atoms, reinitialize=None):    
+        start_time = time.time()
+        # Create the preconditioner:
+        if (self.P is None or self.ctr_of_evals >= self.nbr_of_evals):
+            self._make_sparse_precon(atoms, force_stab=self.force_stab)
+            self.ctr_of_evals = 0
+        self.ctr_of_evals += 1
+        self.logfile.write('--- Precon created in %s seconds ---\n'
+                           % (time.time() - start_time))
+
+
+    def _make_sparse_precon(self, atoms, initial_assembly=False,
+                            force_stab=False):
+        """Create a sparse preconditioner matrix based on the passed atoms.
+
+        Args:
+            atoms: the Atoms object used to create the preconditioner.
+
+        Returns:
+            A scipy.sparse.csr_matrix object, representing a d*N by d*N matrix
+            (where N is the number of atoms, and d is the value of self.dim).
+            BE AWARE that using numpy.dot() with this object will result in
+            errors/incorrect results - use the .dot method directly on the
+            sparse matrix instead.
+
+        """
+        self.logfile.write('creating eam hessian: terms={}'.format(self.terms))
+        start_time = time.time()
+        self.P  = self.calculator.calculate_hessian_matrix(atoms, terms = self.terms)
+        val, vec = eigsh(self.P, k = 1, which= "SA")
+        a_s = 0.00001
+        a = a_s
+        for k in range(100):
```

Comment:
OK, I see how you are ensuring positive definiteness. What I describe in the main thread is maybe a better approach, since it requires only one call to `eigsh()` rather than calling it up to 100 times as you do here. Christoph's approach would be even better.

Reply:
"better" depends very much on the context :) But I do agree that all these eigenvalue decompositions are very expensive and should normally be avoided for preconditioners.